src/agora/txrx/txrx_usrp.cpp

```diff
@@ -0,0 +1,360 @@
+        // receive data
+        // struct Packet* pkt = recv_enqueue_usrp(tid, radio_id, rx_offset);
+        struct Packet* pkt = recv_enqueue_usrp(
+            tid, radio_id, rx_offset, global_frame_id, global_symbol_id);
+
+        // Schedule beacon in the future
+        if (global_symbol_id == 0) {
+            txTimeBs = rxTimeBs
+                + cfg->sampsPerSymbol * cfg->symbol_num_perframe * 20;
+            int tx_ret
+                = radioconfig_->radioTx(0, beaconbuff.data(), 2, txTimeBs);
+            if (tx_ret != (int)cfg->sampsPerSymbol)
+                std::cerr << "BAD Transmit(" << tx_ret << "/"
+                          << cfg->sampsPerSymbol << ") at Time " << txTimeBs
+                          << ", frame count " << global_frame_id << std::endl;
+        }
+
+        if (++radio_id == radio_hi)
+            radio_id = radio_lo;
+
+        // Update global frame_id and symbol_id
+        global_symbol_id++;
+        if (global_symbol_id == (int)cfg->symbol_num_perframe) {
+            global_symbol_id = 0;
+            global_frame_id++;
+        }
+
+        if (pkt == NULL)
+            continue;
+        rx_offset = (rx_offset + cfg->nChannels) % packet_num_in_buffer_;
+
+        if (kIsWorkerTimingEnabled) {
+            int frame_id = pkt->frame_id;
+            // int symbol_id = pkt->symbol_id;
+            if (frame_id > prev_frame_id) {
+                rx_frame_start[frame_id % kNumStatsFrames] = rdtsc();
+                prev_frame_id = frame_id;
+            }
+        }
+    }
+    return 0;
+}
+
+struct Packet* PacketTXRX::recv_enqueue_usrp(
+    int tid, int radio_id, int rx_offset)
+{
+    moodycamel::ProducerToken* local_ptok = rx_ptoks_[tid];
+    char* rx_buffer = (*buffer_)[tid];
+    int* rx_buffer_status = (*buffer_status_)[tid];
+    int packet_length = cfg->packet_length;
+
+    // if rx_buffer is full, exit
+    int nChannels = cfg->nChannels;
+    struct Packet* pkt[nChannels];
+    void* samp[nChannels];
+    for (int ch = 0; ch < nChannels; ++ch) {
+        // if rx_buffer is full, exit
+        if (rx_buffer_status[rx_offset + ch] == 1) {
+            printf("Receive thread %d rx_buffer full, offset: %d\n", tid,
+                rx_offset);
+            cfg->running = false;
+            break;
+        }
+        pkt[ch] = (struct Packet*)&rx_buffer[(rx_offset + ch) * packet_length];
+        samp[ch] = pkt[ch]->data;
+    }
+
+    long long frameTime;
+    if (!cfg->running
+        || radioconfig_->radioRx(radio_id, samp, frameTime) <= 0) {
+        return NULL;
+    }
+
+    int frame_id = (int)(frameTime >> 32);
+    int symbol_id = (int)((frameTime >> 16) & 0xFFFF);
+    int ant_id = radio_id * nChannels;
+    for (int ch = 0; ch < nChannels; ++ch) {
+        new (pkt[ch]) Packet(frame_id, symbol_id, 0 /* cell_id */, ant_id + ch);
+        // move ptr & set status to full
+        // rx_buffer_status[rx_offset + ch] = 1; // has data, after it is read, it is set to 0
```

Comment:
It should be safe to uncomment this line now. BTW, clang-format does not do line break for comments. We need to do it manually when the comment length is too long.

Reply:
done.